PGP 2.0 compatibility updates

dcdenu4 opened the pull request:
This is a large PR that updates InVEST for PGP 2.0 including:

- Moving to GDAL 3 support from GDAL 2
- `pgp.convolve_2d` keyword argument update from `ignore_nodata` to `ignore_nodata_and_edges`
- `pgp.get_[raster|vector]_info` keyword argument update from `projection` to `projection_wkt`
- removing of `pgp.testing` module from testing
- linting InVEST tests

**GDAL Notes:**

The implementation and language used for transforming coordinate systems was taken from Pygeoprocessing. The big change was with models using CoordinateTransformations, which has now been moved to a single entry point in utils.py. The major GDAL 3 differences have been heavily documented elsewhere and discussed at length and for this PR it is mostly the axis ordering that effects InVEST. This has been worked around by defaulting model wide to the traditional ordering, as is done in Pygeoprocessing.

The only tests that were affected were for crop_production. Therefore expected results were updated in invest-test-data repository and PR'd here: https://bitbucket.org/natcap/invest-test-data/pull-requests/11/updating-expected-test-results-for-using/diff

This PR fixes #106
This PR fixes #143 
This PR fixes #102 
This PR fixes #27

dcdenu4 commented:
This PR replaces #106 for GDAL 3.0 incorporation.

dcdenu4 commented:
Ignore the mac test errors for the moment, they appear to be related to a `nump.int64` type not being handled by pgp... because GDAL has no 64 bit int type. Going to explicitly set these to `int32`.

davemfish commented on tests/test_recreation.py:
```diff
@@ -844,8 +842,8 @@ def test_square_grid_regression(self):
 
-        pygeoprocessing.testing.assert_vectors_equal(
-            out_grid_vector_path, expected_grid_vector_path, 1E-6)
+        _assert_vector_attributes_eq(
```
I see this function already existed and was used elsewhere in this file, basically as a means to regression test final model results. But when this function replaces `pygeoprocessing.testing.assert_vectors_equal` then we're no longer asserting equal geometries. Honestly I think that's fine for most of these tests, but this one in particular is testing the `_grid_vector` function, so it is the geometry the needs to be tested.

If we do add a `utils._assert_vectors_equal`, then this could be another use-case.

davemfish commented on tests/test_recreation.py:
```diff
@@ -861,8 +859,8 @@ def test_hex_grid_regression(self):
 
-        pygeoprocessing.testing.assert_vectors_equal(
-            out_grid_vector_path, expected_grid_vector_path, 1E-6)
+        _assert_vector_attributes_eq(
```
Same as above, this happens to be a test where geometries matter.

davemfish commented on tests/test_recreation.py:
```diff
@@ -927,8 +925,8 @@ def test_existing_output_shapefiles(self):
 
-        pygeoprocessing.testing.assert_vectors_equal(
-            out_grid_vector_path, expected_grid_vector_path, 1E-6)
+        _assert_vector_attributes_eq(
```
This test, and some of the previous that I didn't comment on, I think are okay with only asserting on the attribute values since geometry creation is covered by the other tests I commented on.

davemfish commented on tests/test_utils.py:
```diff
@@ -625,3 +654,99 @@ def test_csv_latin_1_encoding(self):
+        x, y, _ = transformer.TransformPoint(lon, lat)
+
+        expected_x = -124.525
+        expected_y = 44.525
```
Looks like we could do without these vars and just use the already defined `lon` and `lat` as the expected data?

davemfish commented on tests/test_utils.py:
```diff
@@ -625,3 +654,99 @@ def test_csv_latin_1_encoding(self):
+        x, y, _ = transformer.TransformPoint(known_x, known_y)
+
+        expected_x = 378816.2531852932
+        expected_y = 4931317.807472325
```
Same comment as above, but now that I see all the tests, I see how it's nice to keep a consistent pattern. Could define `expected_x = known_x`, though?

dcdenu4 replied:
Hey @davemfish , I'm not sure what you were looking for here but I made some changes here 484b2a1b7c33fa2445149cd01b9e90cc85ce4512 that might address the issue?

davemfish replied:
Hey @dcdenu4 , I just meant that these same exact numbers are being assigned to two different variables in this test because the input coords are expected to be identical to the output coords. No big deal though.

phargogh commented on tests/test_wave_energy.py:
```diff
@@ -528,12 +523,12 @@ def test_valuation_suffix(self):
-    def _assert_point_vectors_equal(a_path, b_path):
+    def _assert_point_vectors_equal(a_vector_path, b_vector_path):
         """Assert that two point geometries in the vectors are equal.
 
-        Parameters:
-            a_path (str): a path to an OGR vector.
-            b_path (str): a path to an OGR vector.
+        Args:
+            a_vector_path (str): a path to an OGR vector.
+            b_vector_path (str): a path to an OGR vector.
```
Especially if we end up moving the other vector-testing method to a function in `utils`, I wonder if this assertion method could also be subsumed by it?  At the very least, it'd be another use case.

dcdenu4 replied:
This assertion method is gone.

davemfish commented on tests/test_reporting.py:
```diff
@@ -118,13 +115,16 @@ def test_generate_report_dict_script(self):
+        "skipping due to different number truncation in py36 and py27")
     def test_generate_report_shape_json(self):
         """Reporting: testing full report w/ shape table data and json file."""
         from natcap.invest import reporting
 
         workspace_dir = self.workspace_dir
-        workspace_dir = 'C:/Users/dmf/projects/invest_dev/py36_compatibility/reporting'
+        workspace_dir = os.path.join(
+            'C:', os.sep, 'Users', 'dmf', 'projects', 'invest_dev',
+            'py36_compatibility', 'reporting')
```
@phargogh , good catch. I don't have much memory of this, but it sounds like the expected html differed slightly between python versions.  Given how little use there is of the `reporting` module, I'm very okay with removing the test entirely. We've been living without if for a while anyway by skipping it. Otherwise, I guess we unskip the test and update the regression data so that it passes on python3.

dcdenu4 replied:
I removed it.

dcdenu4 replied:
Removed here: 4961598ed7b354b751e9009dc7864f757be9cedf

dcdenu4 commented on HISTORY.rst:
```diff
@@ -1,15 +1,30 @@
-* SDR's compiled core now defines its own ``SQRT2`` instead of relying on an
-  available standard C library definition.  This new definition helps to avoid
-  some compiler issues on Windows.
+* sdr
```
@davemfish @phargogh it makes sense to me for the model labels to come from `invest list`. That way if model naming changes, like recently happened with SDR, NDR, there's a goto place to get the latest names. I don't love the idea of trying to maintain a static commented list in `History.rst` as models become deprecated, or new ones added... happy to talk about this.

dcdenu4 commented on tests/test_cli.py:
```diff
@@ -149,9 +151,8 @@ def test_model_alias(self):
-        with unittest.mock.patch(
-                'natcap.invest.coastal_blue_carbon.coastal_blue_carbon.execute',
-                return_value=None) as patched_model:
+        targ = 'natcap.invest.coastal_blue_carbon.coastal_blue_carbon.execute'
```
Ha! I think I did that to lazily avoid a line too long

dcdenu4 replied:
Fixed in 395dd4d6348f252778e320b0302712b3854593e0

dcdenu4 commented on tests/test_datastack.py:
```diff
@@ -1,25 +1,78 @@
+
+            expected_geom = feature.GetGeometryRef()
+            expected_geom_wkt = expected_geom.ExportToWkt()
+            actual_geom = feature.GetGeometryRef()
```
Good catch Dave.

dcdenu4 commented:
@phargogh @davemfish would it make more sense to put `_assert_vectors_equal` and other shared testing functions into a new module `tests/testing_utilities.py` instead of `invest/utils.py`? It seems off to me that an assertion function would live in the `utils.py` module. Then I'd feel obligated to write a test for the assertion function in `tests/test_utils.py`. Feel free to call me crazy here.

dcdenu4 commented on tests/test_reporting.py:
```diff
@@ -75,12 +73,11 @@ def test_generate_report_csv_style(self):
-        pygeoprocessing.testing.assert_text_equal(
+        self.assertEqual(
             args['out_uri'],
             os.path.join(
                 REGRESSION_DATA, 'html_reports', 'report_csv_style.html'))
```
Great catch.

dcdenu4 replied:
Fixes here 4961598ed7b354b751e9009dc7864f757be9cedf

dcdenu4 commented on HISTORY.rst:
```diff
@@ -1,15 +1,30 @@
+    * ``get_raster_info`` / ``get_vector_info`` keyword ``projection`` to 
+      ``projection_wkt``.
+    * Updating InVEST tests that were dependent on the removed ``pgp.testing`` 
+      module.
```
Fixed in 7de7c4dfc9a2ac29f7e79b21227c36b8dd6ff5fc

dcdenu4 commented on src/natcap/invest/utils.py:
```diff
@@ -539,3 +547,29 @@ def mean_pixel_size_and_area(pixel_size_tuple):
+            to transform TO
+        osr_axis_mapping_strategy (int): OSR axis mapping strategy for 
+            ``SpatialReference`` objects. Defaults to
+            ``utils.DEFAULT_OSR_AXIS_MAPPING_STRATEGY``. This parameter shoulc
```
Fixed in 48ac546b166c0a962f9bc33861e310e784045ff4

dcdenu4 commented on tests/test_hra.py:
```diff
@@ -424,11 +424,55 @@ def _make_criteria_csv(
+
+            expected_geom = feature.GetGeometryRef()
+            expected_geom_wkt = expected_geom.ExportToWkt()
+            actual_geom = feature.GetGeometryRef()
```
Broke function out to `utils` and fixed in 7645f84866241bef9c54f99a812b4c6c163fe41b

dcdenu4 commented on src/natcap/invest/utils.py:
```diff
@@ -539,3 +547,29 @@ def mean_pixel_size_and_area(pixel_size_tuple):
+
+    """
+    base_ref.SetAxisMappingStrategy(osr_axis_mapping_strategy) 
+    target_ref.SetAxisMappingStrategy(osr_axis_mapping_strategy) 
```
I had never thought about that. To be safe I went ahead and made a transform from a copy of the incoming osr spatial refs 589839aba1751a59bfccc154d1b5c9be5e1a8d0e

dcdenu4 commented on tests/test_datastack.py:
```diff
@@ -1,25 +1,78 @@
+            expected_geom_wkt = expected_geom.ExportToWkt()
+            actual_geom = feature.GetGeometryRef()
+            actual_geom_wkt = actual_geom.ExportToWkt()
+            assert(expected_geom_wkt == actual_geom_wkt)
```
Now using that method in the `utils._assert_vectors_equal` function

dcdenu4 commented on tests/test_hra.py:
```diff
@@ -196,30 +196,25 @@ def _make_raster_from_array(base_array, target_raster_path, projected=True):
-        wrong_layer_type (bool): if true, write a type different from `habitat` or
-            `stressor`.
+        wrong_layer_type (bool): if true, write a type different from
+            `habitat` or `stressor`.
```
Updated here: 4961598ed7b354b751e9009dc7864f757be9cedf

dcdenu4 commented:
Hey @phargogh and @davemfish I think I've addressed all the changes you've brought up. The big ones being:

1. `utils._assert_vectors_equal` has replaced all like calls, including those in `recreation` that were only interested in attribute fields and not geometry.
2. I added a comment block to `HISTORY.rst`, please check it out and provide feedback.
3. Using `shapely` to check geometries
4. Creating CoordinateTransform from copied OSR Spatial References so as not to mutate inputs

I think that covers most of it... and checks pass!

davemfish commented on tests/test_utils.py:
```diff
@@ -743,10 +746,322 @@ def test_projected_to_projected_transformer(self):
+
+        result = utils._assert_vectors_equal(shape_path, shape_copy_path)
+        self.assertFalse(result[0])
+        self.assertTrue("Vector projections are not the same" in result[1])
```
Great tests! I noticed one thing that's not tested is catching geometries that are not equal. What do you think about adding one more? Or even re-purposing one of the existing geometry tests if you want to. I don't think it's super-critical to test different geometry types.

dcdenu4 replied:
Oops! Definitely should check that.

dcdenu4 replied:
Added a test where geometry is expected to fail here: https://github.com/natcap/invest/pull/150/commits/9777b24707ce158d386e4c0403c56eb15ddd2061

dcdenu4 commented on tests/test_hra.py:
```diff
@@ -920,8 +879,8 @@ def test_hra_regression_euclidean_linear(self):
-            HraRegressionTests._assert_vectors_equal(
-                output_vector, expected_vector, precision=6)
+            _assert_vectors_equal(
+                output_vector, expected_vector, tolerance_places=6)
```
Eagle eye

dcdenu4 commented on src/natcap/invest/utils.py:
```diff
@@ -550,26 +556,160 @@ def mean_pixel_size_and_area(pixel_size_tuple):
+        expected_vector_path (string): path on disk to a gdal Vector dataset
+            to use as the ground truth.
+        tolerance_places (int): the decimal tolerance for comparing field
+            attribute values, default=3.
```
Fixed these issues and adding better naming here: https://github.com/natcap/invest/pull/150/commits/6fb2b329d4f14380594ca6de95110f3b4c40774f

phargogh commented on src/natcap/invest/utils.py:
```diff
@@ -539,3 +553,163 @@ def mean_pixel_size_and_area(pixel_size_tuple):
+        actual_feat_count = actual_layer.GetFeatureCount()
+        expected_feat_count = expected_layer.GetFeatureCount()
+        if expected_feat_count != actual_feat_count:
+            return (
+                False,
+                "Vector feature counts are not the same. \n"
+                f"Expected feature count: {expected_feat_count}. \n"
+                f"Actual feature count: {actual_feat_count}. ")
```
Hey @dcdenu4 I'm not sure I'm interpreting this correctly.  It looks like this is returning the validity and error message, which looks inconsistent with the way that it's being used in some of the tests.  In `tests/test_fisheries.py` (line 261), for example (other files use this as well), we have:
```python
utils._assert_vectors_equal(
    aoi_path,
    os.path.join(out_workspace_dir, 'test_aoi_for_subset.shp'))
```

If there were an error in the vector comparison, would the `False` returned be captured and interpreted as a test error?  I know we used to explicitly raise `AssertionError` ... is that no longer required to indicate a test failure?

dcdenu4 replied:
Holy toledos. Good call. Since the 'assertion' function was living inside `utils` I wanted to be able to provide better and more informative failure messages. I thought I had updated all the tests to reflect how the function returned... 

But now that you mention it, perhaps raising `AssertionError` with a message is more descriptive of what the function wants to be.

dcdenu4 replied:
Fixed here https://github.com/natcap/invest/pull/150/commits/ced2e8f4f3cfebcddde923d0a878d8d43ff3b1ae

dcdenu4 commented:
Hey @phargogh and @davemfish I think I've addressed everything to this point. `_assert_vectors_equal` now raises `AssertionError` if vectors are not equal and returns `None` on success. https://github.com/natcap/invest/pull/150/commits/ced2e8f4f3cfebcddde923d0a878d8d43ff3b1ae
